libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -10,157 +10,269 @@
             }
         }
 
-        public async Task<IEnumerable<KeyValuePair<string, object>>> Read(string[] keys)
+        /// <summary>
+        /// Loads store items from storage.
+        /// </summary>
+        /// <param name="keys">Array of item keys to read from the store.</param>
+        public async Task<IEnumerable<KeyValuePair<string, object>>> Read(params string[] keys)
         {
-            var storeItems = new List<KeyValuePair<string, object>>(keys.Length);
-            foreach (string key in keys)
+            if (keys == null || keys.Length == 0)
+            {
+                throw new ArgumentException("Please provide at least one key to read from storage.", nameof(keys));
+            }
+
+            var table = await GetTable().ConfigureAwait(false);
+
+            var readTasks = keys.Select(async key =>
             {
-                var entityKey = GetEntityKey(key);
-                var result = await this.Table.ExecuteAsync(TableOperation.Retrieve<StoreItemEntity>(entityKey.PartitionKey, entityKey.RowKey)).ConfigureAwait(false);
-                if ((HttpStatusCode)result.HttpStatusCode == HttpStatusCode.OK)
+                var ek = new EntityKey(key);
+                var tableEntity = await table.ExecuteAsync(TableOperation.Retrieve<DynamicTableEntity>(ek.PartitionKey, ek.RowKey)).ConfigureAwait(false);
+
+                if (tableEntity.HttpStatusCode == (int)HttpStatusCode.OK)
                 {
-                    var value = ((StoreItemEntity)result.Result).AsObject();
-                    var valueStoreItem = value as IStoreItem;
-                    if (valueStoreItem != null)
-                    {
-                        valueStoreItem.eTag = result.Etag;
-                    }
-                    storeItems.Add(new KeyValuePair<string, object>(key, value));
+                    // re-create expected object
+                    StorageEnvelope envelope = StorageEnvelope.AsStoreItemEntity(tableEntity);
+                    return new KeyValuePair<string, object>(key, envelope.StoreItem);
                 }
-            }
-            return storeItems;
-        }
 
+                return new KeyValuePair<string, object>();
+            });
 
+            return (await Task.WhenAll(readTasks).ConfigureAwait(false))
+                .Where(kv => kv.Key != null);
+        }
+
+        /// <summary>
+        /// Saves store items to storage.
+        /// </summary>
+        /// <param name="changes">Map of items to write to storage.</param>
+        /// <returns></returns>
         public async Task Write(IEnumerable<KeyValuePair<string, object>> changes)
         {
-            foreach (var change in changes)
+            if (changes == null) throw new ArgumentNullException(nameof(changes));
+
+            var storeItems = changes.Select(kv => new StorageEnvelope(new EntityKey(kv.Key), kv.Value));
```

Comment:
So, I didn't catch this before and I apologize for that. I'm not proposing you _have to_ fix what I'm about to say to close out the PR, but I would definitely file a follow up issue to address this. It's a bit of a shame we're having to still spend cycles on supporting ETags/optimistic concurrency since there is already consensus that we want to remove optimistic concurrency. So this issue likely doesn't need to be independently addressed, but rather just closed out when we remove ETags over all.

---

This code is allocating `O(N)` instances of `StorageEnvelope` (which is a class, thus creating GC pressure), one for each change, just so we can have the helper property `ETag` tell us whether or not a _potential_ `ETag` value is present/valid. Also, because the `ETag` property is really just a helper, we're paying `O(2N)` calls to that property, first for the `null` check and then for the `.Length` check, that does a non-trivial amount of work of type checking and conditional property evaluation (this is probably negligible and doesn't allocate, so... take with grain of salt).

Also, far less important, but worth pointing out: because the initial results aren't buffered, in the case that the `.Any()` evaluates to true, we're going to end up allocating another O(N) instances of `StorageEnvelope` when the `IEnumerable<>` is evaluated again with the `.Select` just to project the list of keys.

Honestly, I don't see enough value in `StorageEnvelope` right now to justify its overhead; it doesn't offer much that couldn't be done as just helper methods. Given the way it appears to be used, we could at least consider changing it to a struct to avoid GC overhead.

Reply:
Makes lots of sense. The StorageEnvelope had many more tasks, but not anymore. Dropped in favor a two static helper methods.

Latest changes:
- Remove StorageEnvelope in favor on simpler Helper methods
- Made EntityKey a struct
- Update message for Bogus ETag on entities